### PR TITLE
rockchip: support the suspend/resume for rk3399

### PR DESCRIPTION
**Pull request by rkchrome**
1.Add the suspend/resume for rk3399 SoCS
2.Add cpu clusters suspend for rk3399 SoCS

Change-Id: Id602779016b41d6281f4ba40a20229d909b28e46
Signed-off-by: Shengfei xu <xsf@rock-chips.com>
Signed-off-by: Tony Xie <tony.xie@rock-chips.com>

### CODE REVIEW

File: plat/rockchip/common/aarch64/plat_helpers.S [hunk 1]
```diff
@@ -35,6 +35,9 @@
 #include <cortex_a72.h>
 #include <plat_private.h>
 #include <platform_def.h>
+#if NEED_PMU_MACROS_S
+#include <plat_pmu_macros.S>
+#endif
```
Comment (danh-arm) [inline plat/rockchip/common/aarch64/plat_helpers.S]: minor: Conditionally including this header isn't really necessary - it just clutters the code.

File: plat/rockchip/rk3368/include/platform_def.h [hunk 2]
```diff
@@ -74,6 +74,8 @@
 					 PLATFORM_CLUSTER_COUNT +	\
 					 PLATFORM_CORE_COUNT)
 
+#define PLATFORM_CLST_TO_CPUID_SHIFT	6
```
Comment (danh-arm) [inline plat/rockchip/rk3368/include/platform_def.h]: Prefixing this macro with `PLATFORM_` and placing it in `platform_def.h` implies that this is a platform porting macro when it isn't. Better to just call it `CLST_TO_CPUID_SHIFT` and include in `rk3399_def.h`.

File: plat/rockchip/common/aarch64/plat_helpers.S [hunk 3]
```diff
@@ -192,20 +195,23 @@ endfunc plat_crash_console_putc
...
+#if NEED_PMU_MACROS_S
+	func_rockchip_clst_warmboot
+#endif
 	/* --------------------------------------------------------------------
 	 * big cluster id is 1
 	 * big cores id is from 0-3, little cores id 4-7
 	 * --------------------------------------------------------------------
 	 */
-	add	x0, x1, x0, lsr #6
+	add	x21, x19, x20, lsr #PLATFORM_CLST_TO_CPUID_SHIFT
```
Comment (danh-arm) [inline plat/rockchip/common/aarch64/plat_helpers.S]: minor: Could continue using x0 instead of x21.

File: plat/rockchip/common/aarch64/plat_helpers.S [hunk 4]
```diff
@@ -252,3 +258,6 @@ cpuson_flags:
 	.rept	PLATFORM_CORE_COUNT
 	.word	0
 	.endr
+#if NEED_PMU_MACROS_S
+rockchip_clst_warmboot_data
```
Comment (danh-arm) [inline plat/rockchip/common/aarch64/plat_helpers.S]: minor: indent macro call

File: plat/rockchip/common/drivers/pmu/pmu_com.h [hunk 5]
```diff
@@ -31,7 +31,7 @@
  * Use this macro to instantiate lock before it is used in below
  * rockchip_pd_lock_xxx() macros
  */
-DEFINE_BAKERY_LOCK(rockchip_pd_lock);
+static DEFINE_BAKERY_LOCK(rockchip_pd_lock);
```
Comment (danh-arm) [inline plat/rockchip/common/drivers/pmu/pmu_com.h]: minor: It's a bit strange that this is being declared in the header at all. Better to move this and the functions below into pmu.c, and use `DECLARE_BAKERY_LOCK` here instead.

File: plat/rockchip/common/plat_pm.c [hunk 6]
```diff
@@ -154,14 +143,22 @@ int rockchip_pwr_domain_on(u_register_t mpidr)
  ******************************************************************************/
 void rockchip_pwr_domain_off(const psci_power_state_t *target_state)
 {
-	assert(RK_CORE_PWR_STATE(target_state) == PLAT_MAX_OFF_STATE);
+	if (RK_CORE_PWR_STATE(target_state) != PLAT_MAX_OFF_STATE)
+		return;
```
Comment (danh-arm) [inline plat/rockchip/common/plat_pm.c]: It's a bit suspicious that you've changed an assert to a silent return. Was this condition being fired in certain circumstances or are you just being defensive? In general, asserts are preferred if this really is a programming error.

File: plat/rockchip/common/plat_pm.c [hunk 7]
```diff
@@ -170,17 +167,19 @@ void rockchip_pwr_domain_off(const psci_power_state_t *target_state)
...
 void rockchip_pwr_domain_suspend(const psci_power_state_t *target_state)
 {
-	if (RK_CORE_PWR_STATE(target_state) == PLAT_MAX_RET_STATE)
-		return;
+	uint32_t lvl;
+	plat_local_state_t lvl_state;
 
-	assert(RK_CORE_PWR_STATE(target_state) == PLAT_MAX_OFF_STATE);
+	if (RK_CORE_PWR_STATE(target_state) != PLAT_MAX_OFF_STATE)
+		return;
```
Comment (danh-arm) [inline plat/rockchip/common/plat_pm.c]: Ditto

File: plat/rockchip/rk3399/drivers/pmu/pmu.c [hunk 8]
```diff
@@ -67,6 +67,350 @@ __attribute__ ((section("tzfw_coherent_mem")))
...
+		wait_cnt++;
+		if (!(wait_cnt % 1000)) {
+			INFO("%s:st=%x(%x)\n", __func__,
+			     mmio_read_32(PMU_BASE + PMU_BUS_IDLE_ST),
+			     bus_state);
+			INFO("%s:st=%x(%x)\n", __func__,
+			     mmio_read_32(PMU_BASE + PMU_BUS_IDLE_ACK),
+			     bus_ack);
+		}
+	}
```
Comment (danh-arm) [inline plat/rockchip/rk3399/drivers/pmu/pmu.c]: Potentially could hang here. Consider breaking out of loop after a certain time and cancelling op.

File: plat/rockchip/rk3399/drivers/pmu/pmu.c [hunk 9]
```diff
@@ -249,11 +667,20 @@ static int cores_pwr_domain_on(unsigned long mpidr, uint64_t entrypoint)
 	return 0;
 }
 
-static int cores_pwr_domain_off(void)
+static int cores_pwr_domain_off(uint32_t lvl, plat_local_state_t lvl_state)
```
Comment (danh-arm) [inline plat/rockchip/rk3399/drivers/pmu/pmu.c]: It's confusing that you have 2 different functions for suspend (`cores_pwr_domain_suspend()` and `hlvl_pwr_domain_suspend()`) but 1 combined function for off (`cores_pwr_domain_off()`). Similarly for `cores_pwr_domain_on()`.

File: plat/rockchip/rk3399/drivers/pmu/pmu.c [hunk 10]
```diff
@@ -315,25 +779,77 @@ static void sys_slp_config(void)
...
+		       BIT(PMU_DDRIO1_RET_EN) |
+		       BIT(PMU_DDRIO_RET_HW_DE_REQ) |
+		       BIT(PMU_PLL_PD_EN) |
+		       BIT(PMU_CLK_CENTER_SRC_GATE_EN) |
+		       BIT(PMU_OSC_DIS) |
+		       BIT(PMU_PMU_USE_LF);
+
+	mmio_setbits_32(PMU_BASE + PMU_WKUP_CFG4, BIT(PMU_CLUSTER_L_WKUP_EN));
+	mmio_setbits_32(PMU_BASE + PMU_WKUP_CFG4, BIT(PMU_CLUSTER_B_WKUP_EN));
+	mmio_setbits_32(PMU_BASE + PMU_WKUP_CFG4, BIT(PMU_GPIO_WKUP_EN));
```
Comment (danh-arm) [inline plat/rockchip/rk3399/drivers/pmu/pmu.c]: Looks like a bug fix here that perhaps should be a separate commit.

**Comment by danh-arm** [comment]
jenkins: test this please

File: plat/rockchip/common/aarch64/plat_helpers.S (hunk 1 above)
Comment (rkchrome) [inline plat/rockchip/common/aarch64/plat_helpers.S]: The codes are for all of socs, we consider that maybe some socs is not need，so we add this macros. but we consider it is not necessary now.
Okay remove it.

File: plat/rockchip/rk3368/include/platform_def.h (hunk 2 above)
Comment (rkchrome) [inline plat/rockchip/rk3368/include/platform_def.h]: It is used in plat_helpers.S, all of socs need it for getting cpu id,So we put it here.

File: plat/rockchip/common/aarch64/plat_helpers.S (hunk 3 above)
Comment (rkchrome) [inline plat/rockchip/common/aarch64/plat_helpers.S]: Maybe the value keeping x21 will be used as a parameter with invoking function like x20 in func_rockchip_clst_warmboot。

File: plat/rockchip/common/aarch64/plat_helpers.S (hunk 4 above)
Comment (rkchrome) [inline plat/rockchip/common/aarch64/plat_helpers.S]: Done

File: plat/rockchip/common/drivers/pmu/pmu_com.h (hunk 5 above)
Comment (rkchrome) [inline plat/rockchip/common/drivers/pmu/pmu_com.h]: Done

File: plat/rockchip/common/plat_pm.c (hunk 6 above)
Comment (rkchrome) [inline plat/rockchip/common/plat_pm.c]: Done

File: plat/rockchip/common/plat_pm.c (hunk 7 above)
Comment (rkchrome) [inline plat/rockchip/common/plat_pm.c]: Done

File: plat/rockchip/rk3399/drivers/pmu/pmu.c (hunk 8 above)
Comment (rkchrome) [inline plat/rockchip/rk3399/drivers/pmu/pmu.c]: Done

File: plat/rockchip/rk3399/drivers/pmu/pmu.c (hunk 9 above)
Comment (rkchrome) [inline plat/rockchip/rk3399/drivers/pmu/pmu.c]: The cpu's suspend function may be more difficult than off,So higher level （clusters and system） suspend function is divided. But using divided function is clear。

File: plat/rockchip/rk3399/drivers/pmu/pmu.c (hunk 10 above)
Comment (rkchrome) [inline plat/rockchip/rk3399/drivers/pmu/pmu.c]: It is just a input mistake, we wanted to invoke mmio_setbits_32() here, but we inputed a unexpected function name. So I think it is not necessary to make a separate commit.

File: plat/rockchip/rk3368/include/platform_def.h (hunk 2 above)
Comment (danh-arm) [inline plat/rockchip/rk3368/include/platform_def.h]: But the PLATFORM_ prefix implies this is used by generic code. If you want to define a macro that is used by all Rockchip platforms, then a better prefix would be (for example) PLAT_RK_.

File: plat/rockchip/common/aarch64/plat_helpers.S (hunk 3 above)
Comment (danh-arm) [inline plat/rockchip/common/aarch64/plat_helpers.S]: I don't really understand your comment. This instruction doesn't care what the previous value of x21 (or x0) is. But it's only a minor thing so I won't push it.

**Comment by danh-arm** [comment]
jenkins: retest this please

**Comment by danh-arm** [comment]
We have problems in our CI setup - I will merge this next week. It's up to you if you want to address the minor comments (here or in a future patch).

File: plat/rockchip/rk3368/include/platform_def.h (hunk 2 above)
Comment (rkchrome) [inline plat/rockchip/rk3368/include/platform_def.h]: Okay, done

File: plat/rockchip/common/aarch64/plat_helpers.S (hunk 3 above)
Comment (rkchrome) [inline plat/rockchip/common/aarch64/plat_helpers.S]: Okay. x0 instead of the x21.

**Comment by danh-arm** [comment]
jenkins: retest this please
